packages/imperative/src/events/src/ImperativeEventEmitter.ts

```diff
@@ -0,0 +1,288 @@
+     * Check to see if the Imperative Event Emitter instance has been initialized
+     */
+    private ensureClassInitialized() {
+        if (!ImperativeEventEmitter.initialized) {
+            throw new ImperativeError({msg: "You must initialize the instance before using any of its methods."});
+        }
+    }
+
+    /**
+     * Check to see if the directory exists, otherwise, create it : )
+     * @param directoryPath Zowe or User dir where we will write the events
+     */
+    private ensureEventsDirExists(directoryPath: string) {
+        try {
+            if (!fs.existsSync(directoryPath)) {
+                fs.mkdirSync(directoryPath);
+            }
+        } catch (err) {
+            throw new ImperativeError({ msg: `Unable to create '.events' directory. Path: ${directoryPath}`, causeErrors: err });
+        }
+    }
+
+    /**
+     * Check to see if the file path exists, otherwise, create it : )
+     * @param filePath Zowe or User path where we will write the events
+     */
+    private ensureEventFileExists(filePath: string) {
+        try {
+            if (!fs.existsSync(filePath)) {
+                fs.closeSync(fs.openSync(filePath, 'w'));
+            }
+        } catch (err) {
+            throw new ImperativeError({ msg: `Unable to create event file. Path: ${filePath}`, causeErrors: err });
+        }
+    }
+
+    /**
+     * Helper method to initialize the event
+     * @param eventName The type of event to initialize
+     * @returns The initialized ImperativeEvent
+     */
+    private initEvent(eventName: ImperativeEventType | string): ImperativeEvent {
+        this.ensureClassInitialized();
+        return new ImperativeEvent({ appName: this.appName, eventName, isUser: this.isUserEvent(eventName), logger: this.logger });
+    }
+
+    /**
+     * Helper method to write contents out to disk
+     * @param location directory to write the file (i.e. emit the event)
+     * @param event the event to be written/emitted
+     * @internal We do not want developers writing events directly, they should use the `emit...` methods
+     */
+    private writeEvent(location: string, event: ImperativeEvent) {
+        event.location = location;
+        this.ensureEventsDirExists(location);
+        fs.writeFileSync(join(location, event.type), JSON.stringify(event.toJson(), null, 2));
+    }
+
+    /**
+     * Helper method to create watchers based on event strings and list of callbacks
+     * @param eventName type of event to which we will create a watcher for
+     * @param callbacks list of all callbacks for this watcher
+     * @returns The FSWatcher instance created
+     */
+    private setupWatcher(eventName: string, callbacks: ImperativeEventCallback[] = []): fs.FSWatcher {
+        const dir = this.getEventDir(eventName);
+        this.ensureEventsDirExists(dir);
+        this.ensureEventFileExists(join(dir, eventName));
+
+        const watcher = fs.watch(join(dir, eventName), (event: "rename" | "change") => {
```

Comment:
How do you manage to skip the second optional parameter of fs.watch?
It has a signature like this:
fs.watch( filename[, options][, listener] )

It looks like you only supply the first and third parameters. I am just trying to understand.

Reply:
This is possible thanks to the multiple declarations (a.k.a. [overload signatures](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)) of the fs.watch method.
![image](https://github.com/zowe/zowe-cli/assets/37381190/bfd1a137-e75c-468e-aa6b-e3358a075a8d)

Reply:
Thanks. I looked for another signature, but I did not find one. At least now I know that I am not crazy. 😁